zjit/src/invariants.rs

```diff
@@ -99,3 +103,53 @@ pub fn track_bop_assumption(
         to: side_exit_ptr,
     });
 }
+
+/// Track a patch point for a callable method entry (CME).
+pub fn track_cme_assumption(
+    cme: *const rb_callable_method_entry_t,
+    patch_point_ptr: CodePtr,
+    side_exit_ptr: CodePtr
+) {
+    let invariants = ZJITState::get_invariants();
+    invariants.cme_patch_points.entry(cme).or_default().insert(Jump {
+        from: patch_point_ptr,
+        to: side_exit_ptr,
+    });
+}
+
+/// Called when a method is redefined. Invalidates all JIT code that depends on the CME.
+#[unsafe(no_mangle)]
+pub extern "C" fn rb_zjit_cme_invalidate(cme: *const rb_callable_method_entry_t) {
+    // If ZJIT isn't enabled, do nothing
+    if !zjit_enabled_p() {
+        return;
+    }
+
+    with_vm_lock(src_loc!(), || {
+        let invariants = ZJITState::get_invariants();
+        // Get the CMD's jumps and remove the entry from the map as it has been invalidated
+        if let Some(jumps) = invariants.cme_patch_points.remove(&cme) {
+            let cb = ZJITState::get_code_block();
+            debug!("CME is invalidated: {:?}", cme);
+
+            // Invalidate all patch points for this CME
+            for jump in jumps {
+                cb.with_write_ptr(jump.from, |cb| {
+                    let mut asm = Assembler::new();
+                    asm_comment!(asm, "CME is invalidated: {:?}", cme);
+                    asm.jmp(jump.to.into());
+                    asm.compile(cb).expect("can write existing code");
+                });
+            }
+            cb.mark_all_executable();
+        }
+
+        // Also invalidate the compiled code of the method itself
```

Comment:
Why invalidate the compiled iseq?

Reply:
My implementation is [crashing on Ubuntu](https://github.com/ruby/ruby/actions/runs/16323762653/job/46108291739?pr=13875) and my hypothesis is that it's caused by not cleaning up associated jit pointers in iseq objects. So the change (and the 2nd commit) is to try verifying that.

But now seeing there are [similar build failures on master](https://github.com/ruby/ruby/actions/runs/16325452354/job/46114254086) this may not be the case and I'll drop this change if that's confirmed true.